File: app_java_wiki_chat/spring-backend/src/main/java/com/platform/securelearning/crypto/AesGcmEncryptionService.java
package com.platform.securelearning.crypto;

import java.security.GeneralSecurityException;
import java.security.SecureRandom;
import java.util.Base64;
import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

/**
 * AES-256-GCM at-rest encryption. This is an interim placeholder standing in for the
 * post-quantum symmetric key that cpp-crypto-core's ML-KEM exchange will eventually
 * derive (see {@link EncryptionService}); AES-GCM itself gives no quantum resistance,
 * only confidentiality/integrity for data already stored today.
 */
@Service
public class AesGcmEncryptionService implements EncryptionService {

    private static final Logger log = LoggerFactory.getLogger(AesGcmEncryptionService.class);
    private static final String ALGORITHM = "AES/GCM/NoPadding";
    private static final int GCM_TAG_LENGTH_BITS = 128;
    private static final int NONCE_LENGTH_BYTES = 12;

    private final SecretKey key;
    private final SecureRandom random = new SecureRandom();

    public AesGcmEncryptionService(@Value("${app.crypto.symmetric-key-base64:}") String configuredKeyBase64) {
        this.key = configuredKeyBase64.isBlank() ? generateEphemeralKey() : decodeKey(configuredKeyBase64);
    }

    private SecretKey generateEphemeralKey() {
        log.warn("No app.crypto.symmetric-key-base64 configured — generating an ephemeral "
                + "AES-256 key for this process only. Chat history will be unreadable after "
                + "restart. This is a dev-only placeholder pending the cpp-crypto-core PQC bridge.");
        try {
            KeyGenerator keyGenerator = KeyGenerator.getInstance("AES");
            keyGenerator.init(256);
            return keyGenerator.generateKey();
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("AES key generation is unavailable", e);
        }
    }

    private SecretKey decodeKey(String base64Key) {
        return new SecretKeySpec(Base64.getDecoder().decode(base64Key), "AES");
    }

    @Override
    public EncryptedPayload encrypt(byte[] plaintext) {
        try {
            byte[] nonce = new byte[NONCE_LENGTH_BYTES];
            random.nextBytes(nonce);
            Cipher cipher = Cipher.getInstance(ALGORITHM);
            cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BITS, nonce));
            byte[] ciphertext = cipher.doFinal(plaintext);
            return new EncryptedPayload(ciphertext, nonce);
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("Encryption failed", e);
        }
    }

    @Override
    public byte[] decrypt(EncryptedPayload payload) {
        try {
            Cipher cipher = Cipher.getInstance(ALGORITHM);
            cipher.init(Cipher.DECRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BITS, payload.nonce()));
            return cipher.doFinal(payload.ciphertext());
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("Decryption failed — wrong key or corrupted data", e);
        }
    }
}

File: app02_csharp_wiki_chat/aspnet-backend/SecureLearning.Backend/Crypto/AesGcmEncryptionService.cs
using System.Security.Cryptography;

namespace SecureLearning.Backend.Crypto;

/// <summary>
/// AES-256-GCM at-rest encryption — an interim placeholder standing in for the
/// post-quantum symmetric key cpp-crypto-core's ML-KEM exchange will eventually derive.
/// AES-GCM gives confidentiality/integrity for stored data, not quantum resistance.
/// Key comes from configuration (<c>App:Crypto:SymmetricKeyBase64</c>, env
/// <c>APP_CRYPTO_KEY</c>); when unset a random process-lifetime key is generated,
/// which is fine for dev/tests but makes history unreadable after a restart.
/// </summary>
public sealed class AesGcmEncryptionService : IEncryptionService
{
    private const int NonceLengthBytes = 12;
    private const int TagLengthBytes = 16;
    private const int KeyLengthBytes = 32;

    private readonly byte[] key;

    public AesGcmEncryptionService(string? configuredKeyBase64, ILogger<AesGcmEncryptionService>? logger = null)
    {
        if (string.IsNullOrWhiteSpace(configuredKeyBase64))
        {
            logger?.LogWarning(
                "No App:Crypto:SymmetricKeyBase64 configured — generating an ephemeral AES-256 key "
                + "for this process only. Chat history will be unreadable after restart. This is a "
                + "dev-only placeholder pending the cpp-crypto-core PQC bridge.");
            key = RandomNumberGenerator.GetBytes(KeyLengthBytes);
        }
        else
        {
            key = Convert.FromBase64String(configuredKeyBase64);
        }
    }

    public EncryptedPayload Encrypt(byte[] plaintext)
    {
        byte[] nonce = RandomNumberGenerator.GetBytes(NonceLengthBytes);
        byte[] ciphertextWithTag = new byte[plaintext.Length + TagLengthBytes];
        using var aes = new AesGcm(key, TagLengthBytes);
        // Tag is appended to the ciphertext so EncryptedPayload stays a two-field
        // record matching the Java backend's storage shape.
        aes.Encrypt(nonce, plaintext,
            ciphertextWithTag.AsSpan(0, plaintext.Length),
            ciphertextWithTag.AsSpan(plaintext.Length, TagLengthBytes));
        return new EncryptedPayload(ciphertextWithTag, nonce);
    }

    public byte[] Decrypt(EncryptedPayload payload)
    {
        int plaintextLength = payload.Ciphertext.Length - TagLengthBytes;
        if (plaintextLength < 0)
        {
            throw new CryptographicException("Ciphertext shorter than its authentication tag — corrupted data");
        }
        byte[] plaintext = new byte[plaintextLength];
        using var aes = new AesGcm(key, TagLengthBytes);
        aes.Decrypt(payload.Nonce,
            payload.Ciphertext.AsSpan(0, plaintextLength),
            payload.Ciphertext.AsSpan(plaintextLength, TagLengthBytes),
            plaintext);
        return plaintext;
    }
}
